src/box/lua/net_cluster.lua

```diff
@@ -0,0 +1,466 @@
+    cluster_check(self, 'watch_leaders')
+    local watcher = {
+        _cluster = self,
+        _key = key,
+        _func = func,
+        _opts = opts or {},
+        _states = {},
+    }
+    setmetatable(watcher, cluster_watcher_mt)
+    for _, replicaset in pairs(self._replicasets) do
+        watch_replicaset(watcher, replicaset)
+    end
+    self._watchers[watcher] = watcher
+    return watcher
+end
+
+-- Trigger internal load of config (in case of `config_mode` == 'auto')
+-- set new cluster config (in case of `config_mode` == 'manual')
+-- In the first case `config` must be nil, in the second - it must be
+-- a map replicaset_name -> replicaset_cfg (table).
+-- replicaset_cfg may have members:
+--  connect_cfg: if provided, it will be passed to net.replicaset.connect.
+--   (if not, `replicaset_name` will be passed along with connect_opts).
+--  roles: optional list of roles (map or array) that the replicaset have.
+--   if present, affects filtering by role in some methods.
+function cluster_methods:reload_config(config)
+    cluster_check(self, 'reload_config')
+    if config then
+        if self._cfg.config_mode ~= 'manual' then
+            box.error(box.error.ILLEGAL_PARAMS,
+                    'manual reload expects manual config mode')
+        end
+        reload_config_impl(self, config)
+    else
+        if self._cfg.config_mode ~= 'auto' then
+            box.error(box.error.ILLEGAL_PARAMS,
+                    'force reload expects auto config mode')
+        end
+        reload_config_impl(self, collect_config())
+    end
+end
+
+-- Close all watchers and connections.
+function cluster_methods:close()
+    cluster_check(self, 'close')
+    while true do
+        local watcher = next(self._watchers)
+        if not watcher then
+            break
+        end
+        watcher:unregister()
+    end
+    for _, replicaset in pairs(self._replicasets) do
+        close_replicaset_connection(replicaset)
+    end
+    table.clear(self._replicasets)
+    if self._reload_watcher then
+        self._reload_watcher:unregister()
+        self._reload_watcher = nil
+    end
+    self._is_closed = true
+end
+
+-- get for default cluster (which is created if needed).
+local function get(replicaset_name)
+    if not default_cluster then
+        default_cluster = cluster_new(cluster_cfg_defaults)
+    end
+    return default_cluster:get(replicaset_name)
+end
+
+-- get_connections for default cluster (which is created if needed).
+local function get_connections(opts)
+    if not default_cluster then
+        default_cluster = cluster_new(cluster_cfg_defaults)
+    end
+    return default_cluster:get_connections(opts)
+end
+
+-- watch_leaders for default cluster (which is created if needed).
+local function watch_leaders(key, func, opts)
+    if not default_cluster then
+        default_cluster = cluster_new(cluster_cfg_defaults)
+    end
+    return default_cluster:watch_leaders(key, func, opts)
+end
```

Comment:
I have an idea (inspired by earlier discussions with @Mons) to make the default instace and a user created instace APIs the same.

I mean, if there are the following objects:

```lua
local net_cluster = require('net.cluster')
local my_net_cluster = require('net.cluster').new()
```

The usage is the following now.

```lua
net_cluster.watch_leaders(<...>) -- self is NOT passed
my_net_cluster:watch_leaders(<...>)
```

But I would like the following one.

```lua
net_cluster:watch_leaders(<...>) -- self is passed
my_net_cluster:watch_leaders(<...>)
```

I see two pros:

* The same API: we can implement a function that accepts any of these objects and just call the methods.
* The default instance has methods that explicitly accept a state in the `self` argument instead of implicit passing it via captured upvalues (that feels hacky).

A possible con is that it is inconsistent with some of existing modules: for example, `json`. OTOH, I use the proposed approach in new modules, where appropriate: for example, `config`.

I don't insist here in any way, but if you like the idea too, I see no reasons to discard it.